### PR TITLE
Make ``__all__`` reflect public (and not) API

### PR DESCRIPTION
**Pull request by nstarman**
Focus on the top-level namespace. Now ``__all__`` is the definition what is and isn't public API.

PEP8 (https://peps.python.org/pep-0008/#public-and-internal-interfaces) says every module should have an `__all__`, even if it's empty!

TODO:
- [ ] Should ``flrw.__all__`` have stuff, as in should ``cosmology.flrw.FlatLambdaCDM`` be public, or just ``cosmology.FlatLambdaCDM``?
- [x] Changelog? I suppose under the assumption that ``__all__`` has always been the definition of public API and this module was just untidy, there should be a changelog.

Signed-off-by: Nathaniel Starkman (@nstarman) <nstarkman@protonmail.com>

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [ ] Do the proposed changes actually accomplish desired goals?
- [ ] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [ ] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [ ] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [ ] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [ ] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label. Codestyle issues can be fixed by the [bot](https://docs.astropy.org/en/latest/development/workflow/development_workflow.html#pre-commit).
- [ ] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label. If this is a manual backport, use the `skip-changelog-checks` label unless special changelog handling is necessary.
- [ ] Is this a big PR that makes a "What's new?" entry worthwhile and if so, is (1) a "what's new" entry included in this PR and (2) the "whatsnew-needed" label applied?
- [ ] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [ ] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

### CODE REVIEW

File: astropy/cosmology/tests/test_public_api.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    a bit difficult for developers to know what is and isn't public in a specific file,
+    without opening the module's ``__init__.py``. To compensate, we have a dunder
+    attribute ``__public_api__`` in modules where ``__all__`` is not defined. This test
+    checks that the ``__public_api__`` is a subset of the ``__all__`` in the parent
+    module.
+    """
+    cosmology_dir = Path(__file__).parent.parent
+
+    for pth in sorted(cosmology_dir.glob("**/*.py")):
+        for pth in sorted(cosmology_dir.rglob("*.py")):
```
Comment (nstarman) [inline astropy/cosmology/tests/test_public_api.py:19]: What happened here?!?

Comment (nstarman) [reply astropy/cosmology/tests/test_public_api.py]: Fixed.

**Comment by nstarman** [comment]
ping @WilliamJamieson @mhvk for your thoughts on this move to clean up the API definitions.

**Comment by mhvk** [comment]
Sorry, teaching has started so I won't have much time for review and will focus on units, etc. 

But FWIW, I still don't like empty `__all__` -- readability counts and if `__all__` is present, it should be there to help developers, and the only use I can see is to tell what in a given module is supposed to be used elsewhere, i.e., is not private *to the module* (rather than to the (sub)project) -- this provides a good way to avoid having to use leading underscores in many names (or, worse, in module names). Generally, in astropy I think we have used it mostly to enable `import *` in `__init__`, which is consistent with this. 

Anyway, I'm not keen to make this an astropy-wide policy, but if for cosmology you prefer it, that's fine.

File: astropy/cosmology/__init__.py [hunk 2]
```diff
@@ -8,27 +8,40 @@
...
+    "WMAP9",  # In `__getattr__`
+    "Planck13",  # In `__getattr__`
+    "Planck15",  # In `__getattr__`
+    "Planck18",  # In `__getattr__`
+    # funcs
+    "z_at_value",
+    "cosmology_equal",
+    # public modules
+    "realizations",
+    "units",
```
Comment (WilliamJamieson) [inline astropy/cosmology/__init__.py:42]: Could this be built out of the `__public_api__` in the other modules?

Comment (nstarman) [reply astropy/cosmology/__init__.py:42]: Yes, except ``__all__`` is special-cased in mypy and static type-checkers and I don't think dynamically building it from ``__public_api__`` will be compatible.

File: astropy/cosmology/flrw/base.py [hunk 3]
```diff
@@ -33,7 +33,8 @@ def quad(*args, **kwargs):
         raise ModuleNotFoundError("No module named 'scipy.integrate'")
 
 
-__all__ = ["FLRW", "FlatFLRWMixin"]
+__all__ = []  # No public API -- see cosmology/__init__.py
+__public_api__ = ["FLRW", "FlatFLRWMixin"]
```
[the remark this answers is not in the document]
Comment (nstarman) [reply astropy/cosmology/flrw/base.py:37]: We don't want users to import it from ``astropy.cosmology.flrw.base``. The specific file is an implementation detail. The top-level namespace ``astropy.cosmology`` (and possibly ``astropy.cosmology.flrw``)  is meant to be a static and safe path, like how ``numpy`` puts everything at the top namespace and is free to refactor specific function locations.

[the remark this answers is not in the document]
Comment (saimn) [reply astropy/cosmology/flrw/base.py:37]: But `__all__` doesn't define what's public or not, i.e. `__all__ = []` does not prevent to import directly from `astropy.cosmology.flrw.base`. This change seems a bit confusing to me.

**Comment by saimn** [comment]
> Now __all__ is the definition what is and isn't public API.

But it was already isn't it ?

Python does not have a well defined concept of "public API". I think the general consensus about what's public or not is just that public API is what's present in the documentation (not only for astropy, true also for most big packages, numpy scipy etc.) . 
https://docs.astropy.org/en/latest/cosmology/index.html#module-astropy.cosmology lists all the classes functions etc that can be imported from `astropy.cosmology`, and `dir(astropy.cosmology)` should give the same list, and that's it. 
I don't think introducing a `__public_api__` variable helps, nobody will look at it (unless maybe it becomes a standard for astropy), and it doesn't prevent importing from sub-sub-packages. The only way to have truly public and private API would be to make sub-sub-packages private, but prefixing every module with _ is annoying.

File: astropy/cosmology/flrw/base.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (WilliamJamieson) [reply astropy/cosmology/flrw/base.py:37]: I agree with @saimn, this change is puzzling to me too.

If you really wanted to control the API more fully, you should make the classes and/or modules "private" with a leading underscore.

**Comment by nstarman** [comment]
This PR is based on https://github.com/astropy/astropy/issues/13910. Both Python and SciPy have detailed discussion of how to define and communicate a library's public API  ([Python's reference](https://github.com/python/typing/blob/master/docs/source/libraries.rst#how-much-of-my-library-needs-types) and [scipy's reference](https://docs.scipy.org/doc/scipy/reference/index.html#importing-from-scipy).
In short, as scipy puts it "if a module or package defines ``__all__,`` that authoritatively defines the public interface."

I think adopting this definition is very useful for developers as each file can communicates **exactly** what is or isn't public. Consider the current state

```
cosmology/
    __init__.py:: __all__ = ["Cosmology"]  # Cosmology is public
    core.py:: __all__ = ["Cosmology"]
```

This says that ``from astropy.cosmology.core import Cosmology`` is equally valid to ``from astropy.cosmology import Cosmology``! We don't want this!
This explains why I've added so many empty ``__all__``. 

```
cosmology/
    __init__.py:: __all__ = ["Cosmology"]  # Cosmology is public
    core.py:: __all__ = []
```

Now only ``from astropy.cosmology import Cosmology`` is publicly scoped.
Yes, one can still do ``from astropy.cosmology.core import Cosmology``, but this is clearly marked as private.

To address some Qs:

> But FWIW, I still don't like empty `__all__` -- readability counts and if `__all__` is present, it should be there to help developers.

I think this does help the developer. If ``__all__`` is present and empty then we are communicating that the entire contents of the file are private — no leading underscores necessary. 

Quoting from the scipy docs:
>> Reading the above guidelines one could draw the conclusion that every private module or object starts with an underscore. This is not the case; the presence of underscores do mark something as private, but the absence of underscores do not mark something as public.


With empty ``__all__`` we can define objects in files and ensure that only the module-level import is publicly scoped.

> Generally, in astropy I think we have used it mostly to enable `import *` in `__init__`, which is consistent with this.

An empty ``__all__`` does too and IMO does an even better job protecting public vs private definition locations.


> But it was already isn't it ?

Not really. In my above example one would think that ``astropy.cosmology.core.Cosmology`` was a publicly-scoped means to import the Cosmology class. It is not.

> If you really wanted to control the API more fully, you should make the classes and/or modules "private" with a leading underscore.

That is one method to communicate public vs private. However both [Python](https://github.com/python/typing/blob/master/docs/source/libraries.rst#library-interface-public-and-private-symbols) and [SciPy](https://docs.scipy.org/doc/scipy/reference/index.html#importing-from-scipy) say that ``__all__`` supercedes the leading underscore convention. Quoting from Python:

>> A module can expose an ``__all__`` symbol at the module level that provides a list of names that are considered part of the interface. This overrides all other rules above, allowing imported symbols or symbols whose names begin with an underscore to be included in the interface.

**Comment by nstarman** [comment]
I added the ``__public_api__`` to communicate within a file where nothing is publicly scoped in **that** file that some of the contents are publicly scoped **elsewhere**. Perhaps this is overkill, or not the best means to communicate this information. I'm very open to suggestions for how best to communicate that an object is publicly-scope-elsewhere-but-not-here.

File: astropy/cosmology/flrw/base.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (nstarman) [reply astropy/cosmology/flrw/base.py:37]: See https://github.com/astropy/astropy/pull/14239#issuecomment-1382899762 for full explanation, but by Python guidance ``__all__`` supercedes leading underscores. We can use full and empty ``__all__`` to fully control the API.

**Comment by nstarman** [comment]
Thanks @WilliamJamieson for pointing me to https://peps.python.org/pep-0008/#public-and-internal-interfaces.
PEP8 itself says every module should have an `__all__`, even if it's empty!

**Comment by WilliamJamieson** [comment]
> Thanks @WilliamJamieson for pointing me to [peps.python.org/pep-0008/#public-and-internal-interfaces](https://peps.python.org/pep-0008/#public-and-internal-interfaces). PEP8 itself says every module should have an `__all__`, even if it's empty!

It also says that modules that are internal should be prefixed with a `_`. Maybe you should move all your stuff into modules like `_base.py` and then just have a `base.py` which imports it and deprecates it so that in astropy 6 you can have your total public/private API.

**Comment by nstarman** [comment]
I agree. I would prefer to rename files in a separate PR because of supporting YAML save paths and stuff. This PR is more atomic, focusing just on `__all__`.

File: astropy/cosmology/__init__.py (hunk 2 above)
Comment (nstarman) [reply astropy/cosmology/__init__.py:42]: I've remove `__public_api__` from this PR.

**Comment by WilliamJamieson** [comment]
> I agree on your suggestion for v6.0. I would prefer to rename files in a separate PR because of supporting YAML save paths and stuff. This PR is more atomic, focusing just on `__all__`.

Maybe that should be merged first? It would make the changes here make more sense.

**Comment by nstarman** [comment]
IMO either order is fine. The rest of Astropy does not have leading underscore module names, but uses `__all__` and emphasizes the top-level namespace. This change feels consistent with current Astropy.

**Comment by nstarman** [comment]
Hunh. Started on the module renaming. Maybe that should be done simultaneously with `__all__`  in this PR.

**Comment by nstarman** [comment]
An open question is whether `cosmology.flrw` should be considered private, as a module. All the contents are and have been in the top-level namespace. Should the `cosmology.flrw` be public as well? 
Thoughts @mhvk @WilliamJamieson @eerovaher

**Comment by mhvk** [comment]
I think if the contents are in the top namespace, then there would generally be no need to import from `flwr`, so "private" in that sense.

As said, I'm still not sold on using `__all__ = []` to indicate that, though! I think it is fine if it indicates which parts are imported elsewhere and thus public in that sense.

**Comment by saimn** [comment]
> Thanks @WilliamJamieson for pointing me to https://peps.python.org/pep-0008/#public-and-internal-interfaces.
PEP8 itself says every module should have an __all__, even if it's empty!

Well before that it says:
> Documented interfaces are considered public, unless the documentation explicitly declares them to be provisional or internal interfaces exempt from the usual backwards compatibility guarantees. All undocumented interfaces should be assumed to be internal.

Only `astropy.cosmology` is documented in https://docs.astropy.org/en/latest/cosmology/index.html so `cosmology.flrw` and all other submodules are private.

**Comment by pllim** [comment]
Would this depend on the "public API" APE that is forthcoming? Or can this go in regardless?

**Pull request closed without merging**